### DISCUSSION TITLE
Add chord display to score, renders chord patterns directly above each bar of tab staff.

### DISCUSSION THREAD

**Discussion by vivachenx** [CoderLine/alphaTab#703] [Ideas]
![image](https://user-images.githubusercontent.com/12762833/150262941-bcdaab78-86f8-4b08-a2a0-6b9f68e642d4.png)

Like this picture, Renders chord patterns directly above each bar of tab staff.

Have we considered adding such a rendering method in subsequent versions?

**Comment by vivachenx** [comment]
![image](https://user-images.githubusercontent.com/12762833/150263235-1c3ef9b5-c2b1-4378-916a-ac9fd445f3ff.png)

**Reply by Produj** [reply]
May I ask how it was implemented? Thank you very much

**Comment by Danielku15** [comment]
Hi there. 
While it is not directly planned already to be shipped in a dedicated version, I definitely have it on my "known unsupported features" list. 
When I implemented it on the top, I made a quick PoC for inline chords and in extreme cases it rather caused troubles in the layouting: https://github.com/CoderLine/alphaTab/issues/110#issuecomment-277526500 

Theoretically it should be fairly quick to add them again if we want to ignore the extreme cases and live with overlaps in such cases. It would be just a matter of adjusting the [`ChordEffectInfo`](https://github.com/CoderLine/alphaTab/blob/master/src/rendering/effects/ChordsEffectInfo.ts) to create a  [`ChordDiagramGlyph`](https://github.com/CoderLine/alphaTab/blob/master/src/rendering/glyphs/ChordDiagramGlyph.ts) based on some settings coming from Guitar Pro and/or the some external user input whether to enforce them. Guitar Pro has a setting in the stylesheet which we would respect. 
![image](https://user-images.githubusercontent.com/674916/150408962-2604e8e5-9fbb-412e-9abd-7a2d5d685aea.png)

**Reply by vivachenx** [reply]
Hi there.
Thank you so much for your reply, I just briefly tried adding this feature. I found that if I directly replace the TextGlyph with the ChordDiagramGlyph instance, the rendering is a bit strange.
Then I found another problem, we currently don't seem to support rendering tempo related information
This is alpha result:
![image](https://user-images.githubusercontent.com/12762833/158959660-0d6a8374-4e7e-4c25-985d-5dd54445d044.png)
This is guitar pro result:
<img width="730" alt="image" src="https://user-images.githubusercontent.com/12762833/158960583-131cc9d1-681d-4071-a229-aaf422afa996.png">

**Reply by vivachenx** [reply]
this is my test code:
`import { Beat } from '@src/model/Beat';
import { TextAlign } from '@src/platform/ICanvas';
import { BarRendererBase } from '@src/rendering/BarRendererBase';
import { EffectBarGlyphSizing } from '@src/rendering/EffectBarGlyphSizing';
import { EffectGlyph } from '@src/rendering/glyphs/EffectGlyph';
import { TextGlyph } from '@src/rendering/glyphs/TextGlyph';
import { ChordDiagramGlyph } from '@src/rendering/glyphs/ChordDiagramGlyph';
import { EffectBarRendererInfo } from '@src/rendering/EffectBarRendererInfo';
import { Settings } from '@src/Settings';
import { NotationElement } from '@src/NotationSettings';
import { ChordDiagramContainerGlyph } from '@src/rendering/glyphs/ChordDiagramContainerGlyph';

export class ChordsEffectInfo extends EffectBarRendererInfo {
    public get notationElement(): NotationElement {
        return NotationElement.EffectChordNames;
    }

    public get hideOnMultiTrack(): boolean {
        return false;
    }

    public get canShareBand(): boolean {
        return true;
    }

    public get sizingMode(): EffectBarGlyphSizing {
        return EffectBarGlyphSizing.SingleOnBeat;
    }

    public shouldCreateGlyph(settings: Settings, beat: Beat): boolean {
        return beat.hasChord;
    }

    public createNewGlyph(renderer: BarRendererBase, beat: Beat): EffectGlyph {
        const chordDiagram: ChordDiagramGlyph = new ChordDiagramGlyph(0, 0, beat!.chord);
        return chordDiagram
        // return new TextGlyph(0, 0, beat.chord!.name, renderer.resources.effectFont, TextAlign.Center);
    }

    public canExpand(from: Beat, to: Beat): boolean {
        return false;
    }
}
`
